crates/ruma-client-api/src/discovery/discover_homeserver.rs

```diff
@@ -112,3 +131,163 @@ impl TileServerInfo {
         Self { map_style_url }
     }
 }
+
+/// Information about a specific MatrixRTC focus.
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+#[serde(tag = "type")]
+pub enum RtcFocusInfo {
+    /// A LiveKit RTC focus.
+    #[serde(rename = "livekit")]
+    LiveKit(LiveKitRtcFocusInfo),
+
+    /// A custom RTC focus.
+    #[doc(hidden)]
+    #[serde(untagged)]
+    _Custom(CustomRtcFocusInfo),
+}
+
+#[cfg(feature = "unstable-msc4143")]
+impl RtcFocusInfo {
+    /// A constructor to create a LiveKit RTC focus.
+    pub fn live_kit(service_url: String) -> Self {
+        Self::LiveKit(LiveKitRtcFocusInfo { service_url })
+    }
+
+    /// A constructor to create a custom RTC focus.
+    pub fn custom(focus_type: String, data: JsonObject) -> Self {
+        Self::_Custom(CustomRtcFocusInfo { focus_type, data })
+    }
+
+    /// Returns a reference to the focus type of this RTC focus.
+    pub fn focus_type(&self) -> &str {
+        match self {
+            Self::LiveKit(_) => "livekit",
+            Self::_Custom(custom) => &custom.focus_type,
+        }
+    }
+
+    /// Returns the associated data.
+    ///
+    /// The returned JSON object won't contain the `focus_type` field, please use
+    /// [`.focus_type()`][Self::focus_type] to access that.
+    ///
+    /// Prefer to use the public variants of `RtcFocusInfo` where possible; this method is meant to
+    /// be used for custom focus types only.
+    pub fn data(&self) -> Cow<'_, JsonObject> {
+        fn serialize<T: Serialize>(object: &T) -> JsonObject {
+            match serde_json::to_value(object).expect("rtc focus type serialization to succeed") {
+                JsonValue::Object(object) => object,
+                _ => panic!("all rtc focus types must serialize to objects"),
+            }
+        }
+
+        match self {
+            Self::LiveKit(info) => Cow::Owned(serialize(info)),
+            Self::_Custom(info) => Cow::Borrowed(&info.data),
+        }
+    }
+}
+
+/// Information about a LiveKit RTC focus.
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct LiveKitRtcFocusInfo {
+    /// The service URL for the LiveKit service.
+    #[serde(rename = "livekit_service_url")]
+    pub service_url: String,
+}
+
+/// Information about a custom RTC focus type.
+#[doc(hidden)]
+#[cfg(feature = "unstable-msc4143")]
+#[derive(Clone, Debug, Deserialize, Serialize)]
+#[cfg_attr(not(ruma_unstable_exhaustive_types), non_exhaustive)]
+pub struct CustomRtcFocusInfo {
+    /// The type of RTC focus.
+    #[serde(rename = "type")]
+    pub focus_type: String,
+
+    /// Remaining RTC focus data.
+    #[serde(flatten)]
+    pub data: JsonObject,
+}
+
+#[cfg(test)]
+mod tests {
+    #[cfg(feature = "unstable-msc4143")]
+    mod rtc_foci {
```

Comment:
We don't really need this extra module, you can just put everything directly in `tests`, with `#[cfg(all(test, feature = "unstable-msc4143"))]`

Reply:
You don't need to, you just need to set it on `mod tests`.

Reply:
But that seemed mean for the next person who comes along and adds some tests to this file 🙂